File: src/main/java/com/sunset/lavafishing/util/LavaArmorMaterials.java
package com.sunset.lavafishing.util;

import com.sunset.lavafishing.util.RegistryCollection.ItemCollection;
import com.teammetallurgy.aquaculture.api.AquaArmorMaterials;
import net.minecraft.Util;
import net.minecraft.sounds.SoundEvent;
import net.minecraft.sounds.SoundEvents;
import net.minecraft.util.LazyLoadedValue;
import net.minecraft.util.StringRepresentable;
import net.minecraft.world.item.ArmorItem;
import net.minecraft.world.item.ArmorMaterial;
import net.minecraft.world.item.crafting.Ingredient;

import javax.annotation.Nonnull;
import java.util.EnumMap;
import java.util.function.Supplier;

public enum LavaArmorMaterials implements StringRepresentable, ArmorMaterial
{
    PROMETHIUM("promethium", 38, Util.make(new EnumMap<>(ArmorItem.Type.class), (map) ->
    {
        map.put(ArmorItem.Type.BOOTS, 4);
        map.put(ArmorItem.Type.LEGGINGS, 7);
        map.put(ArmorItem.Type.CHESTPLATE, 9);
        map.put(ArmorItem.Type.HELMET, 4);
    }), 14, SoundEvents.ARMOR_EQUIP_DIAMOND, 2.5F, 0.1F, () -> Ingredient.of(ItemCollection.PROMETHIUM_INGOT.get()));


    public static final StringRepresentable.EnumCodec<AquaArmorMaterials> CODEC = StringRepresentable.fromEnum(AquaArmorMaterials::values);
    private static final EnumMap<ArmorItem.Type, Integer> HEALTH_FUNCTION_FOR_TYPE = Util.make(new EnumMap<>(ArmorItem.Type.class), (map) -> {
        map.put(ArmorItem.Type.BOOTS, 13);
        map.put(ArmorItem.Type.LEGGINGS, 15);
        map.put(ArmorItem.Type.CHESTPLATE, 16);
        map.put(ArmorItem.Type.HELMET, 11);
    });
    private final String name;
    private final int durabilityMultiplier;
    private final EnumMap<ArmorItem.Type, Integer> protectionFunctionForType;
    private final int enchantmentValue;
    private final SoundEvent sound;
    private final float toughness;
    private final float knockbackResistance;
    private final LazyLoadedValue<Ingredient> repairIngredient;

    private LavaArmorMaterials(String name, int durabilityMultiplier, EnumMap<ArmorItem.Type, Integer> protectionFunctionForType, int enchantmentValue, SoundEvent sound, float toughness, float knockbackResistance, Supplier<Ingredient> repairIngredient) {
        this.name = name;
        this.durabilityMultiplier = durabilityMultiplier;
        this.protectionFunctionForType = protectionFunctionForType;
        this.enchantmentValue = enchantmentValue;
        this.sound = sound;
        this.toughness = toughness;
        this.knockbackResistance = knockbackResistance;
        this.repairIngredient = new LazyLoadedValue<>(repairIngredient);
    }

    @Override
    public int getDurabilityForType(@Nonnull ArmorItem.Type armorItemType) {
        return HEALTH_FUNCTION_FOR_TYPE.get(armorItemType) * this.durabilityMultiplier;
    }

    @Override
    public int getDefenseForType(@Nonnull ArmorItem.Type armorItemType) {
        return this.protectionFunctionForType.get(armorItemType);
    }

    @Override
    public int getEnchantmentValue() {
        return this.enchantmentValue;
    }

    @Override
    @Nonnull
    public SoundEvent getEquipSound() {
        return this.sound;
    }

    @Override
    @Nonnull
    public Ingredient getRepairIngredient() {
        return this.repairIngredient.get();
    }

    @Override
    @Nonnull
    public String getName() {
        return this.name;
    }

    @Override
    public float getToughness() {
        return this.toughness;
    }

    @Override
    public float getKnockbackResistance() {
        return this.knockbackResistance;
    }

    @Override
    @Nonnull
    public String getSerializedName() {
        return this.name;
    }
}

File: src/main/java/club/redux/sunset/lavafishing/util/LavaArmorMaterials.kt
package club.redux.sunset.lavafishing.util

import club.redux.sunset.lavafishing.util.RegistryCollection.ItemCollection
import net.minecraft.sounds.SoundEvent
import net.minecraft.sounds.SoundEvents
import net.minecraft.util.StringRepresentable
import net.minecraft.world.item.ArmorItem
import net.minecraft.world.item.ArmorMaterial
import net.minecraft.world.item.crafting.Ingredient
import java.util.*

enum class LavaArmorMaterials(
    private val materialName: String,
    private val durabilityMultiplier: Int,
    private val defenseForType: EnumMap<ArmorItem.Type, Int>,
    private val durabilityForType: EnumMap<ArmorItem.Type, Int>,
    private val enchantmentValue: Int,
    private val sound: SoundEvent,
    private val toughness: Float,
    private val knockbackResistance: Float,
    private val repairIngredient: Ingredient,
) : StringRepresentable, ArmorMaterial {

    PROMETHIUM(
        "promethium",
        38,
        EnumMap<ArmorItem.Type, Int>(ArmorItem.Type::class.java).apply {
            put(ArmorItem.Type.BOOTS, 4)
            put(ArmorItem.Type.LEGGINGS, 7)
            put(ArmorItem.Type.CHESTPLATE, 9)
            put(ArmorItem.Type.HELMET, 4)
        },
        EnumMap<ArmorItem.Type, Int>(ArmorItem.Type::class.java).apply {
            put(ArmorItem.Type.BOOTS, 4)
            put(ArmorItem.Type.LEGGINGS, 7)
            put(ArmorItem.Type.CHESTPLATE, 9)
            put(ArmorItem.Type.HELMET, 4)
        },
        14,
        SoundEvents.ARMOR_EQUIP_DIAMOND,
        2.5f,
        0.1f,
        Ingredient.of(ItemCollection.PROMETHIUM_INGOT.get())
    );

    override fun getDurabilityForType(type: ArmorItem.Type): Int = durabilityForType[type]!! * this.durabilityMultiplier
    override fun getDefenseForType(type: ArmorItem.Type): Int = defenseForType[type]!!
    override fun getEnchantmentValue(): Int = this.enchantmentValue
    override fun getEquipSound(): SoundEvent = this.sound
    override fun getRepairIngredient(): Ingredient = repairIngredient
    override fun getName(): String = this.materialName
    override fun getToughness(): Float = this.toughness
    override fun getKnockbackResistance(): Float = this.knockbackResistance
    override fun getSerializedName(): String = this.materialName
}